From 4a70aed5a0a69f93bb1a950f1b5a3f0fd4cc52c5 Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:36:50 -0400
Subject: [ethernet] i40e/i40evf: Add EEE LPI stats

Message-id: <1409658042-22689-50-git-send-email-sassmann@redhat.com>
Patchwork-id: 90243
O-Subject: [RHEL7.1 PATCH 049/281] i40e/i40evf: Add EEE LPI stats
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Anjali Singhai Jain <anjali.singhai@intel.com>

Add 4 new stats to keep track of EEE LPI (Low Power Idle) state.

Change-ID: Id6316619bb0559789770288b694a54d17f8fac5c
Signed-off-by: Anjali Singhai Jain <anjali.singhai@intel.com>
Acked-by: Shannon Nelson <shannon.nelson@intel.com>
Signed-off-by: Kevin Scott <kevin.c.scott@intel.com>
Signed-off-by: Catherine Sullivan <catherine.sullivan@intel.com>
Tested-by: Kavindya Deegala <kavindya.s.deegala@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit bee5af7e21939c54538b8d1555b4c568e9fbd235)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_ethtool.c b/drivers/net/ethernet/intel/i40e/i40e_ethtool.c
index 8ee224f..d5d5c35 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_ethtool.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_ethtool.c
@@ -114,6 +114,11 @@ static struct i40e_stats i40e_gstrings_stats[] = {
  I40E_PF_STAT("VF_admin_queue_requests", vf_aq_requests),
  I40E_PF_STAT("tx_hwtstamp_timeouts", tx_hwtstamp_timeouts),
  I40E_PF_STAT("rx_hwtstamp_cleared", rx_hwtstamp_cleared),
+ /* LPI stats */
+ I40E_PF_STAT("tx_lpi_status", stats.tx_lpi_status),
+ I40E_PF_STAT("rx_lpi_status", stats.rx_lpi_status),
+ I40E_PF_STAT("tx_lpi_count", stats.tx_lpi_count),
+ I40E_PF_STAT("rx_lpi_count", stats.rx_lpi_count),
 };
 
 #define I40E_QUEUE_STATS_LEN(n) \
diff --git a/drivers/net/ethernet/intel/i40e/i40e_main.c b/drivers/net/ethernet/intel/i40e/i40e_main.c
index 1f10ff2..4791bb3 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_main.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_main.c
@@ -740,6 +740,7 @@ void i40e_update_stats(struct i40e_vsi *vsi)
  u32 rx_page, rx_buf;
  u64 rx_p, rx_b;
  u64 tx_p, tx_b;
+ u32 val;
  int i;
  u16 q;
 
@@ -972,6 +973,20 @@ void i40e_update_stats(struct i40e_vsi *vsi)
   i40e_stat_update32(hw, I40E_GLPRT_RJC(hw->port),
        pf->stat_offsets_loaded,
        &osd->rx_jabber, &nsd->rx_jabber);
+
+  val = rd32(hw, I40E_PRTPM_EEE_STAT);
+  nsd->tx_lpi_status =
+          (val & I40E_PRTPM_EEE_STAT_TX_LPI_STATUS_MASK) >>
+    I40E_PRTPM_EEE_STAT_TX_LPI_STATUS_SHIFT;
+  nsd->rx_lpi_status =
+          (val & I40E_PRTPM_EEE_STAT_RX_LPI_STATUS_MASK) >>
+    I40E_PRTPM_EEE_STAT_RX_LPI_STATUS_SHIFT;
+  i40e_stat_update32(hw, I40E_PRTPM_TLPIC,
+       pf->stat_offsets_loaded,
+       &osd->tx_lpi_count, &nsd->tx_lpi_count);
+  i40e_stat_update32(hw, I40E_PRTPM_RLPIC,
+       pf->stat_offsets_loaded,
+       &osd->rx_lpi_count, &nsd->rx_lpi_count);
  }
 
  pf->stat_offsets_loaded = true;
diff --git a/drivers/net/ethernet/intel/i40e/i40e_type.h b/drivers/net/ethernet/intel/i40e/i40e_type.h
index 5c902f4..d2f0b95 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_type.h
+++ b/drivers/net/ethernet/intel/i40e/i40e_type.h
@@ -1014,6 +1014,11 @@ struct i40e_hw_port_stats {
  u64 tx_size_big;  /* ptc9522 */
  u64 mac_short_packet_dropped; /* mspdc */
  u64 checksum_error;  /* xec */
+ /* EEE LPI */
+ bool tx_lpi_status;
+ bool rx_lpi_status;
+ u64 tx_lpi_count;  /* etlpic */
+ u64 rx_lpi_count;  /* erlpic */
 };
 
 /* Checksum and Shadow RAM pointers */
-- 
1.7.1